### PR TITLE
Fix clang sanitizer invocation

### LINKED ISSUE #1716: json::parse() ubsan regression with v3.7.0
- What is the issue you have?

When upgrading from v3.6.1 to v3.7.0, I did see [a crash in the CI of a library I maintain](https://github.com/measurement-kit/measurement-kit/pull/1887#issuecomment-523041315) when using the undefined behaviour sanitiser with GCC. I was able to cut down complexity and produce a minimal test case that shows the issue is around `json::parse()` (see below).

- Please describe the steps to reproduce the issue. Can you provide a small but working code example?

1. `docker run -it debian:testing-20190812`; then inside the container do:
2. `apt update`
3. `apt install -y g++ nvi wget`
4. `wget https://github.com/nlohmann/json/releases/download/v3.7.0/json.hpp`
5. edit `main.cpp` to contain the code shown below
6. `g++ -fsanitize=undefined -fno-sanitize-recover main.cpp`
7. `./a.out`

where:

```C++
// main.cpp
#include "json.hpp"
int main() {
  std::string m = "";
  auto doc = nlohmann::json::parse(m);
}
```

- What is the expected behavior?

This is what I obtain with nlohmann/json@v3.6.1 as well as with v3.7.1 when (1) not using the sanitiser or (2) `m` is a `const char *`:

```
# ./a.out
terminate called after throwing an instance of 'nlohmann::detail::parse_error'
  what():  [json.exception.parse_error.101] parse error at line 1, column 1: syntax
    error while parsing value - unexpected end of input; expected '[', '{', or a literal
Aborted
```

- And what is the actual behavior instead?

```
# ./a.out
/usr/include/c++/8/ext/new_allocator.h:136:4: runtime error: null pointer passed as
    argument 2, which is declared to never be null
```

See the bottom of this issue for a stacktrace.

- Which compiler and operating system are you using? Is it a [supported compiler](https://github.com/nlohmann/json#supported-compilers)?

Compiler:

```
# g++ --version
g++ (Debian 8.3.0-19) 8.3.0
[snip]
```

System: `debian:testing-20190812`. Upon arriving in this point of the bug report, I also attempted to reproduce the issue with the `debian:stable` image, with `g++ (Debian 8.3.0-6) 8.3.0`, where the behaviour was exactly the same.

- Did you use a released version of the library or the version from the `develop` branch?

I did test v3.6.1, v3.7.0, and a015b78e. Only v0.3.6.1 was working as intended.

- If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?

I don't think this question applies to my case.

- - -

From a `debian:stable` docker container, and with v3.7.0, I obtained this stack trace (which has been edited for wrapping long lines):

```
# head -n 10 json.hpp 
/*
    __ _____ _____ _____
 __|  |   __|     |   | |  JSON for Modern C++
|  |  |__   |  |  | | | |  version 3.7.0
|_____|_____|_____|_|___|  https://github.com/nlohmann/json

Licensed under the MIT License <http://opensource.org/licenses/MIT>.
SPDX-License-Identifier: MIT
Copyright (c) 2013-2019 Niels Lohmann <http://nlohmann.me>.

# cat main.cpp
// main.cpp
#include "json.hpp"
int main() {
  std::string m = "";
  auto doc = nlohmann::json::parse(m);
}

# g++ -g -fno-omit-frame-pointer -fsanitize=undefined -fno-sanitize-recover main.cpp

# export UBSAN_OPTIONS=print_stacktrace=1

# ./a.out 
/usr/include/c++/8/ext/new_allocator.h:136:4: runtime error: null pointer
  passed as argument 2, which is declared to never be null

    #0 0x55820c2f8650 in void __gnu_cxx::new_allocator<
      nlohmann::detail::input_buffer_adapter>::construct<
        nlohmann::detail::input_buffer_adapter, decltype(nullptr),
          unsigned long const&>(nlohmann::detail::input_buffer_adapter*,
            decltype(nullptr)&&, unsigned long const&)
              /usr/include/c++/8/ext/new_allocator.h:136

    #1 0x55820c2f5c04 in void std::allocator_traits<
      std::allocator<nlohmann::detail::input_buffer_adapter> >::
        construct<nlohmann::detail::input_buffer_adapter, decltype(nullptr),
          unsigned long const&>(std::allocator<nlohmann::detail::input_buffer_adapter>&,
            nlohmann::detail::input_buffer_adapter*, decltype(nullptr)&&,
              unsigned long const&) /usr/include/c++/8/bits/alloc_traits.h:475

    #2 0x55820c2f0986 in std::_Sp_counted_ptr_inplace<
      nlohmann::detail::input_buffer_adapter, std::allocator<
        nlohmann::detail::input_buffer_adapter>, (__gnu_cxx::_Lock_policy)2>::
          _Sp_counted_ptr_inplace<decltype(nullptr),
            unsigned long const&>(std::allocator<nlohmann::detail::input_buffer_adapter>,
              decltype(nullptr)&&, unsigned long const&)
                /usr/include/c++/8/bits/shared_ptr_base.h:545

    #3 0x55820c2e71b9 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::
      __shared_count<nlohmann::detail::input_buffer_adapter,
        std::allocator<nlohmann::detail::input_buffer_adapter>, decltype(nullptr),
          unsigned long const&>(nlohmann::detail::input_buffer_adapter*&,
            std::_Sp_alloc_shared_tag<std::allocator<
              nlohmann::detail::input_buffer_adapter> >, decltype(nullptr)&&,
                unsigned long const&) /usr/include/c++/8/bits/shared_ptr_base.h:677

    #4 0x55820c2de847 in std::__shared_ptr<
      nlohmann::detail::input_buffer_adapter, (__gnu_cxx::_Lock_policy)2>::
        __shared_ptr<std::allocator<nlohmann::detail::input_buffer_adapter>,
          decltype(nullptr), unsigned long const&>(
            std::_Sp_alloc_shared_tag<std::allocator<nlohmann::detail::
              input_buffer_adapter> >, decltype(nullptr)&&,
                unsigned long const&) /usr/include/c++/8/bits/shared_ptr_base.h:1342

    #5 0x55820c2c9296 in std::shared_ptr<nlohmann::detail::input_buffer_adapter
      >::shared_ptr<std::allocator<nlohmann::detail::input_buffer_adapter>,
        decltype(nullptr), unsigned long const&>(
          std::_Sp_alloc_shared_tag<std::allocator<
            nlohmann::detail::input_buffer_adapter> >, decltype(nullptr)&&,
              unsigned long const&) /usr/include/c++/8/bits/shared_ptr.h:359

    #6 0x55820c2c3461 in std::shared_ptr<nlohmann::detail::input_buffer_adapter>
      std::allocate_shared<nlohmann::detail::input_buffer_adapter,
        std::allocator<nlohmann::detail::input_buffer_adapter>, decltype(nullptr),
          unsigned long const&>(std::allocator<nlohmann::detail::input_buffer_adapter
            > const&, decltype(nullptr)&&, unsigned long const&) 
              /usr/include/c++/8/bits/shared_ptr.h:706

    #7 0x55820c2b637f in std::shared_ptr<nlohmann::detail::input_buffer_adapter
       > std::make_shared<nlohmann::detail::input_buffer_adapter,
         decltype(nullptr), unsigned long const&>(decltype(nullptr)&&,
           unsigned long const&) /usr/include/c++/8/bits/shared_ptr.h:722

    #8 0x55820c2b579f in nlohmann::detail::input_adapter::input_adapter<
      __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<
       char, std::char_traits<char>, std::allocator<char> > >, 0>
         (__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<
           char, std::char_traits<char>, std::allocator<char> > >,
             __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<
               char, std::char_traits<char>, std::allocator<char> > >) //json.hpp:4166

    #9 0x55820c2b22b4 in nlohmann::detail::input_adapter::input_adapter<
      std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, 0>
        (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >
          const&) //json.hpp:4181

    #10 0x55820c2ab686 in main //main.cpp:5
```

### PR DESCRIPTION
**Pull request by t-b**
The clang version used for the sanitizer got upgraded. In addition we now don't recover from sanitizer issues, thus make passing them a requirement, and also print out the stacktrace on sanitizer issues.

Close #1716.

### CODE REVIEW

**Comment by t-b** [comment]
I think the annotations indizes are sometimes off by one:

```
git grep -h -A2 JSON_HEDLEY_NON_NULL
```

gives

```
...
--
nlohmann/json.hpp:3885:    JSON_HEDLEY_NON_NULL(2)
nlohmann/json.hpp-3886-    input_buffer_adapter(const char* b, const std::size_t l) noexcept
nlohmann/json.hpp-3887-        : cursor(b), limit(b + l)
--
...
```

and in some cases the function even does not have two parameters

```
--
    JSON_HEDLEY_NON_NULL(2)
    bool sax_parse_internal(SAX* sax)
    {
--
```

**Comment by t-b** [comment]
Progress:

- Fixed some of the violations
- Left over ones are in the tests (signed char vs uint8_t)
- I do get an error here with libstd++ from gcc 6.3 with std::vector<bool> which AFAIU is not our bug.

I'm attaching the log here 
[output.txt](https://github.com/nlohmann/json/files/3560712/output.txt).

Suggestions on how to proceed are appreciated.

File: include/nlohmann/detail/conversions/to_json.hpp [hunk 1]
```diff
@@ -153,8 +153,11 @@ struct external_constructor<value_t::array>
     {
         j.m_type = value_t::array;
         j.m_value = value_t::array;
-        j.m_value.array->resize(arr.size());
-        std::copy(std::begin(arr), std::end(arr), j.m_value.array->begin());
+        if (arr.size() > 0)
+        {
+            j.m_value.array->resize(arr.size());
+            std::copy(std::begin(arr), std::end(arr), j.m_value.array->begin());
+        }
```
Comment (jaredgrubb) [inline include/nlohmann/detail/conversions/to_json.hpp]: What was the bug here?

Comment (t-b) [reply include/nlohmann/detail/conversions/to_json.hpp]: 5d83e5e4e425606cb5b6ec33677a6229e0ff5c71 has the details.

Comment (jaredgrubb) [reply include/nlohmann/detail/conversions/to_json.hpp]: Ok, I can buy that the `begin` call should be guarded, but should the `resize` also be guarded? 

I'm worried that skipping the `resize(0)` could have an undesired side effect?

Comment (t-b) [reply include/nlohmann/detail/conversions/to_json.hpp]: @jaredgrubb

I'll move the resize out of the if, make sense.

**Comment by t-b** [comment]
- Cleaned up some commits
- Added UBSAN suppression support and ignore the std::vector<bool> 
see 5da932f for details

**Comment by t-b** [comment]
@jaredgrubb @nlohmann The unicode test with sanitizer timed out, but no errors from the sanitizer.

File: include/nlohmann/detail/input/input_adapters.hpp [hunk 2]
```diff
@@ -147,6 +146,7 @@ class input_buffer_adapter : public input_adapter_protocol
     {
         if (JSON_HEDLEY_LIKELY(cursor < limit))
         {
+            assert(cursor != nullptr && limit != nullptr);
```
Comment (nlohmann) [inline include/nlohmann/detail/input/input_adapters.hpp]: This library uses alternative operator keywords. Please change to:

```suggestion
            assert(cursor != nullptr and limit != nullptr);
```

**Comment by nlohmann** [comment]
> @jaredgrubb @nlohmann The unicode test with sanitizer timed out, but no errors from the sanitizer.

I think we should increase the timeout to make sure the CI can run successfully.

File: include/nlohmann/detail/input/input_adapters.hpp (hunk 2 above)
Comment (t-b) [reply include/nlohmann/detail/input/input_adapters.hpp]: @nlohmann Done. That's something I have to get used to.

**Comment by t-b** [comment]
> I think we should increase the timeout to make sure the CI can run successfully.

@nlohmann Done.

**Comment by t-b** [comment]
Thanks for the review btw.

**Comment by nlohmann** [comment]
* * *

### 🔖 Release item

This issue/PR will be part of the next release of the library. This template helps preparing the release notes.

#### Type

- [ ] ✨ New Feature
- [ ] 🐛 Bug Fix
- [ ] ⚡️ Improvement
- [x] 🔨 Further Change
- [ ] 🔥 Deprecated function

#### Description

- Updated sanitizers and make sanitizer check a mandatory CI step.

* * *

**Comment by nlohmann** [comment]
Thanks a lot!!

**Comment by bassosimone** [comment]
Thanks @t-b and @nlohmann!

**Pull request merged**